woocommerce/woocommerce-ios
# [Subscriptions in Order Creation 3] Update product cards

Closes: #12437
Apologies for the large PR, more than half belongs to additional Unit Tests 🙇 

## Description
As continuation of https://github.com/woocommerce/woocommerce-ios/pull/12447, this PR updates the product collapsible cards in order creation, and adds support for product subscriptions, this includes:
- Simple subscriptions, variable subscriptions and its variations
- Pricing and frequency
- Signup fees, and free trial periods

## Changes
- We add a new property to the `CollapsibleRowViewModel` that contains subscription settings extracted from product meta data for a Subscription-type Product, if any.
- UI-wise we update the collapsible product cards to render this new information when needed, we add billing price, interval, and subscription conditions like free trials or sign up fees. 
- Due to the API accepting values like empty or zero for these, we also perform some validation when rendering those fields. If empty, nil, or zero, we won't render them.

## Caveats
In order to keep this PR under 500 lines we have not handled 2 cases yet: 
- When signup fees are zero, you will still see them in the product selector, but not in the product card (they shouldn't appear in neither). This will be handled here: https://github.com/woocommerce/woocommerce-ios/issues/12471
- When a subscription is on sale, this makes the price and the subscription price to be different. This will be tackled on https://github.com/woocommerce/woocommerce-ios/issues/12516 since needs a design review first.

## Testing instructions
- On a store using WooCommerce Subscriptions, create a simple subscription, a variable subscription, and variations with different prices, billing, sign up fees and/or free trial conditions. For easiness you can use the testing site https://indiemelon.mystagingwebsite.com/ , that already has a bunch of subscriptions setup.
- Go to Orders > `+` > Search for "virtual sub" and add different products to an order (eg: sign-up fee, no sign-up fee, free trial, no free trial, etc, ... (for easiness, just select all of them).
- Add them to the Order
- Observe that collapsed rows render subscriptions conditions appropriately. Eg: No conditions for no signup fee/trial, conditions for signup fee/trial, when interval is 1 we word it as a singular, when is more than 1 we word it as plural, etc, ... 
- Expand the product card and observe that conditions also render when needed.

| Select subscription products | Collapsed conditions | Expanded conditions |
|--------|--------|--------|
| ![simulator_screenshot_5E3C08B6-DBC3-4557-B9A4-1156524A778D](https://github.com/woocommerce/woocommerce-ios/assets/3812076/a20301b8-ad46-4759-9ee0-7d805c01fba9) | ![simulator_screenshot_EE3E56BD-17EE-48AA-B784-33224860E6B2](https://github.com/woocommerce/woocommerce-ios/assets/3812076/ff6db21f-e774-4201-b85e-c24b4a788cba) | ![simulator_screenshot_29D147C2-AA42-468A-B8A7-05FC9AA5B2B2](https://github.com/woocommerce/woocommerce-ios/assets/3812076/08d70eff-f0a0-4942-bafd-241eceb86afa)
